packages/vitest/src/integrations/wait.ts

```diff
@@ -0,0 +1,93 @@
+import { getSafeTimers } from '@vitest/utils'
+
+// The waitFor function was inspired by https://github.com/testing-library/web-testing-library/pull/2
+
+export type WaitForCallback<T> = () => T | Promise<T>
+
+export interface WaitForOptions {
+  /**
+   * @description Time in ms between each check callback
+   * @default 50ms
+   */
+  interval?: number
+  /**
+   * @description Time in ms after which the throw a timeout error
+   * @default 1000ms
+   */
+  timeout?: number
+}
+
+function copyStackTrace(target: Error, source: Error) {
+  if (source.stack !== undefined)
+    target.stack = source.stack.replace(source.message, target.message)
+  return target
+}
+
+export function waitFor<T>(callback: WaitForCallback<T>, options: number | WaitForOptions = {}) {
+  const { setTimeout, setInterval, clearTimeout, clearInterval } = getSafeTimers()
+  const { interval = 50, timeout = 1000 } = typeof options === 'number' ? { timeout: options } : options
+  const STACK_TRACE_ERROR = new Error('STACK_TRACE_ERROR')
+
+  return new Promise<T>((resolve, reject) => {
+    let lastError: unknown
+    let promiseStatus: 'idle' | 'pending' | 'resolved' | 'rejected' = 'idle'
+    let timeoutId: ReturnType<typeof setTimeout>
+    let intervalId: ReturnType<typeof setInterval>
+
+    const onResolve = (result: T) => {
+      if (timeoutId)
+        clearTimeout(timeoutId)
+      if (intervalId)
+        clearInterval(intervalId)
+
+      resolve(result)
+    }
+
+    const handleTimeout = () => {
+      let error = lastError
+      if (!error)
+        error = copyStackTrace(new Error('Timed out in waitFor!'), STACK_TRACE_ERROR)
```

Comment:
One thing I've seen some developers running into is that `waitFor` is failing for them when `useFakeTimers()` is used and they are unaware of this. It could be that they simply forgot to call `useRealTimers()` or that timer mocking was done by some utility function which hides it. 

Could we add some notification here if `waitFor` fails and timers are mocked? E.g. `Timed out in waitFor. ${timersMocked ? 'Note that timers were mocked with useFakeTimers(). This may affect your waitFor usage.' : ''}` or similar. Goal of this would be to provide additional information for debugging failures.

Reply:
Function is using safe timers, so I'm not sure if we need this?

Reply:
`waitFor` itself is running safely and correctly, but if a developer is waiting for a change while timers are mocked, that change won't happen.

Reply:
I prefer to add some notifications here. mocked time may complicate

Reply:
And what should users do then? I don't understand how this is helpful instead of doing the expected? (advance timers)

Reply:
The way it's handled now is what's expected, we need to do nothing here. Add notification just a friendly reminder and help users how to debug the test.

Reply:
How do you fix it tho? Users can't control the timer while `waitFor` is running.

Reply:
Ahh, yes, waitFor will be pause until timed out